components/rollup_node/p2p_interface.md

```diff
@@ -0,0 +1,206 @@
+
+## P2P configuration
+
+### Identification
+
+Nodes have a **separate** network- and consensus-identity.
+The network identity is a `secp256k1` key, used for both discovery and active LibP2P connections.
+
+Common representations of network identity:
+- `PeerID`: a LibP2P specific ID derived from the pubkey (through protobuf encoding, typing and hashing)
+- `NodeID`: a Discv5 specific ID derived from the pubkey (through hashing, used in the DHT)
+- `Multi-address`: an unsigned address, containing: IP, TCP port, PeerID
+- `ENR`: a signed record used for discovery, containing: IP, TCP port, UDP port, signature (pubkey can be derived) and L2 network identification. Generally encoded in base64.
+
+### Discv5
+
+#### Structure
+
+The Ethereum Node Record (ENR) for an Optimism rollup node must contain keyed values:
+
+- An IPv4 address (`ip` field) and/or IPv6 address (`ip6` field).
+- A TCP port (`tcp` field) representing the local libp2p listening port.
+- A UDP port (`udp` field) representing the local discv5 listening port.
+- An Optimism (`optimism` field) L2 network identifier
+
+The `optimism` value is encoded as the concatenation of:
+- chain ID (`varint`)
+- fork ID (`varint`)
+
+Note that DiscV5 is a shared DHT: the L1 consensus and execution nodes, as well as testnet nodes,
+and even external IOT nodes, all communicate records in this large common DHT.
+
+The discovery process in Optimism is a pipeline of node records:
+1. Fill the table with `FINDNODES` if necessary
+2. Pull additional records with searches to random Node IDs if necessary
+3. Pull records from the DiscV5 module when looking for peers
+4. Check if the record contains the `optimism` entry, verify it matches the chain ID and current or future fork number
+5. If not already connected, and not recently disconnected or put on deny-list, attempt to dial.
+
+### LibP2P
+
+#### Transport
+
+TCP transport. Additional transports are supported by LibP2P, but not required.
+
+#### Dialing
+
+Nodes should be publicly dialable, not rely on relay extensions, and able to dial both IPv4 and IPv6.
+
+#### NAT
+
+The listening endpoint must be publicly facing, but may be configured behind a NAT.
+LibP2P has optional automation for this.
+
+#### Peergating and pruning
+
+TODO: max-peers, grace periods, pruning
+
+#### Transport security
+
+[Libp2p-noise][libp2p-noise], `XX` handshake, with the the `secp256k1` P2P identity, as popularized in Eth2.
+
+#### Protocol negotiation
+
+[Multistream-select 1.0][multistream-select] (`/multistream/1.0.0`) is an interactive protocol
+used to negotiate sub-protocols supported in LibP2P peers. Multistream-select 2.0 may be used in the future.
+
+#### Identify
+
+LibP2P offers a minimal identification module to share client version and language.
+This is optional and can be disabled for enhanced privacy.
+It also includes the same protocol information, which can speed up initial connections.
```

Comment:
Not sure what that last sentence means.

Are there other advantages in sharing client version & language? (btw language? programming language??)

Reply:
Programming language yes. If you work with multiple p2p implementations, it narrows down bugs and weird behavior: very useful debug information. Similarly for the client-version, we can identify which version of the code was affected when we find a bug in external nodes that we have otherwise not much insight into.

Reply:
I see, thanks. Why is it "the same protocol information" vs "protocol version information" or something like that?